Add an examples explanation contribution page (#1949)
bevyengine/bevy-website: content/learn/contribute/helping-out/explaining-examples.md

+++
title = "Explaining Examples"
insert_anchor_links = "right"
[extra]
weight = 6
+++

Writing up an explanation for an example can be helpful for providing context and **telling a story** of how problems are solved by combining different features of Bevy.

## Where to write explanations

Example explanations should be **in the .rs file of the example**. This keeps them directly tied to the example's code, which will keep eyes on the explanation when changes are made to the code itself.

## Style guide

Everything in [Writing Documentation](../writing-docs) also applies here, with some context-specific tweaks as the context for examples is different.

{% callout(type="warning") %}
There is no predictable order to how people visit the examples pages, people will look for what they need in the moment and then go to it. No page can assume that another page has been read first, or at least if it does it may need to link directly to that other page.
{% end %}

### Building a narrative and avoiding rote repetition

> A subject should be identified for an example so that the narrative can be built around it. If someone is writing an explanation on an example that is only "Setting up the camera" then there's plenty of space to give to the details of setting up the camera and all the possible things one could draw on and why, but we would not want to overemphasise extra things we put in the scene to show that the camera is in fact working. Similarly, for a UI example we will want to explain through a narrative of why we want to achieve making a UI this way within a common, abstracted context for building a UI in games.

Reviewer: ```suggestion
A subject should be identified for an example so that the narrative can be built around it. If someone is writing an explanation on an example that is only about "Setting up the camera" then there's plenty of space to give to the details of setting up the camera and all the possible things one could draw on and why. But we would not want to overemphasise extra things we put in the scene to show that the camera is in fact working. Similarly, for a UI example we will want to explain through a narrative of why we want to achieve making a UI this way within a common, abstracted context for building a UI in games.
```
One of the main sentences was pretty long, I split it in two.

Also: 
- "all the possible things one could draw on and why": I feel like a more concrete example would be better/simpler than this abstract designation here
- same thing for "within a common, abstracted context for building a UI in games" ?

Feel free to ignore my remarks

Author: It's all good, my prose isn't hammered in for this context 😅

### Building a narrative

A subject should be identified for an example so that the narrative can be built around it. If someone is writing an explanation on an example that is only "Setting up the camera" then there's plenty of space to give to the details of setting up the camera and all the possible things one could draw on, but this can't be copied-and-pasted into another example. Paragraphs should **not** be written in a context-free way to be copy-pasted between examples, this is a [non-goal](#non-goals-of-explaining-examples).

### What's in a paragraph

Paragraphs should ideally cover **1 concept/topic** before moving on and be fairly short (generally 50 to 150 words, this is not a hard rule and neither minimising or maximising is "better"). This is to benefit **skim-reading** as a way of finding information quickly and then being able to properly contextualise it once you've found something that resembles what you're looking for.

Subjects of paragraphs should usually be name dropped at the beginning of those paragraphs, again this serves the purpose of skim-reading by having there be a semi-reliable area where a concept's name is dropped.

### Who looks at examples

> This is slightly different set of demographic assumptions than in [Writing Documentation](../writing-docs#learning-material-structure) as we're making the assumption people are already at the example pages and have some more nuance available to us with our assumptions.

Reviewer: I feel like the sentence is a bit contrived. We could remove it as what follows is pretty self-explanatory.
Or we could change it to be more concise
```suggestion
The demographic assumptions are a bit different than in [Writing Documentation](../writing-docs#learning-material-structure) as we are assuming that people are already navigating the example pages.
```

> There's 3 kinds of people we're aiming to write this kind of documentation for, with their own reasons they're coming to the Bevy example pages:

Reviewer: ```suggestion
There are 3 kinds of people we're aiming to write this kind of documentation for, each with their own reasons to visit the Bevy example pages:
```

### Rust experience

Experience with Rust should be assumed to be intermediate. If something complex is happening with Rust we shouldn't be afraid to explain it, but it should be brief and in the context of a specific Bevy feature. Rust doesn't need to be over-explained.

## Questions to ask when planning an explanation

1. What are the key concepts in the example?
2. Whose eye might be caught by the example?
    1. Imagine the assumptions one would make when looking at an example and what they would want to get out of it, across skill sets and amount of experience.
3. Is the topic an "advanced" feature or something plug-and-play?
    1. Does the technology need to be explained, or just how to use it?
    2. Something that requires an already existing deep knowledge of another domain may only need to nod towards that domain, not fully explain it.
4. Does the explanation you're writing also line up with the [general documentation style guide](../writing-docs)?

> ## Non-goals of explaining examples

Reviewer: I would propose merging the `## Non-goals of explaining examples` paragraph with `### Building a narrative` (or bringing them together in `style guide`). Both of the paragraphs contents are touching a similar subject, and, when reading top to bottom, the first one directs you to the second one almost immediately, but is the only one to link to this.

Author: Agreed, and followed up.
